src/plugins/data/server/search/es_search/es_search_rxjs_helpers.ts

```diff
@@ -0,0 +1,162 @@
+import type { TransportRequestPromise } from '@elastic/elasticsearch/lib/Transport';
+
+import { shimAbortSignal } from './shim_abort_signal';
+import { toSnakeCase } from './to_snake_case';
+import { getDefaultSearchParams } from './get_default_search_params';
+import { getTotalLoaded } from './get_total_loaded';
+import { isCompleteResponse, ISearchOptions } from '../../../common/search/es_search';
+
+import type { IKibanaSearchRequest, IKibanaSearchResponse } from '../../../common/search/types';
+import type { AsyncOptions } from './get_default_search_params';
+import type { SearchUsage } from '../collectors';
+
+export type KibanaSearchParams = Record<string, any>;
+
+export interface EsRawResponse extends SearchResponse<any> {
+  id?: string;
+  is_partial?: boolean;
+  is_running?: boolean;
+}
+
+type MapArgsFn = (
+  params: KibanaSearchParams,
+  config: SharedGlobalConfig
+) => SearchArgs | Promise<SearchArgs>;
+
+type SearchMethod = (
+  params: any,
+  options?: any
+) => TransportRequestPromise<ApiResponse<EsRawResponse>>;
+
+export interface SearchArgs {
+  params: KibanaSearchParams;
+  options?: Record<string, any>;
+}
+
+export const getSearchArgs = (uiSettingsClient: IUiSettingsClient, mapArgsFn: MapArgsFn) =>
+  mergeMap(async (config: SharedGlobalConfig) => {
+    const params = {
+      ...(await getDefaultSearchParams(uiSettingsClient)),
+    };
+
+    return mapArgsFn ? await mapArgsFn(params, config) : { params };
+  });
+
+export const doSearch = (
+  searchClient: SearchMethod,
+  abortSignal?: AbortSignal,
+  usage?: SearchUsage
+) => ({ params, options }: SearchArgs) =>
+  from(
+    new Promise<EsRawResponse>(async (resolve, reject) => {
+      try {
+        const apiResponse = await shimAbortSignal(
+          searchClient(toSnakeCase(params), options && toSnakeCase(options)),
+          abortSignal
+        );
+        const rawResponse = (apiResponse as ApiResponse<EsRawResponse>).body;
+
+        usage?.trackSuccess(rawResponse.took);
+
+        resolve(rawResponse);
+      } catch (e) {
+        usage?.trackError();
+        reject(e);
+      }
+    })
+  );
+
+export const doPartialSearch = (
+  searchClient: SearchMethod,
+  partialSearchСlient: SearchMethod,
+  requestId: IKibanaSearchRequest['id'],
+  asyncOptions: AsyncOptions,
+  { abortSignal, waitForCompletion }: ISearchOptions,
+  usage?: SearchUsage
+) => ({ params, options }: SearchArgs) => {
+  const isCompleted = (response: EsRawResponse) =>
+    !Boolean(response.is_partial || response.is_running);
+
+  const partialSearch = (id: string): Observable<EsRawResponse> =>
+    from(
+      partialSearchСlient(
+        {
+          id,
+          ...toSnakeCase(asyncOptions),
+        },
+        options
+      )
+    ).pipe(
+      switchMap(({ body }: ApiResponse<EsRawResponse>) => {
+        return waitForCompletion && !isCompleted(body) && body.id
+          ? timer(1000).pipe(switchMap(() => partialSearch(body.id!)))
```

Comment:
One thing I'm not a big fan of is that, although this logic is similar to the polling code we have in data enhanced `search_interceptor.ts`, we're not reusing any of it. Is there enough of an overlap to warrant writing a utility and moving to `common/`?

Reply:
Also, the `waitForCompletion` parameter... does it do anything other than what `first()` would accomplish? Could we get away with using `first()` everywhere instead of introducing this new parameter?